tests/unit/general/test_mdspan.cpp

```diff
@@ -0,0 +1,337 @@
+   FiniteElementSpace fes(&mesh, &fec);
+   const int ND = fes.GetNDofs();
+
+   SECTION("EGDA, left")
+   {
+      MDGridFunction<4> gsa(NE, NG, &fes, NA);
+      const int gsa_0123 = gsa.Offset(0, 1, 2, 3);
+      REQUIRE(gsa_0123 ==  0 + 1*(NE) + 2*(NE*NG) + 3*(NE*NG*ND));
+      mfem::Array<int> vdofs(ND);
+      gsa.GetVDofs(0, 1, vdofs, 3);
+      REQUIRE(0 <= vdofs.Min());
+      REQUIRE(vdofs[2] == gsa_0123);
+      REQUIRE(vdofs.Max() < NE*NG*NA*ND);
+
+   }
+
+   SECTION("EGDA, right")
+   {
+      MDGridFunction<4, MDLayoutRight<4>> gsa(NE, NG, &fes, NA);
+      const int gsa_0123 = gsa.Offset(0,1,2,3);
+      REQUIRE(gsa_0123 == 0*(NG*ND*NA) + 1*(ND*NA) + 2*(NA) + 3);
+      mfem::Array<int> vdofs(ND);
+      gsa.GetVDofs(0, 1, vdofs, 3);
+      REQUIRE(0 <= vdofs.Min());
+      REQUIRE(vdofs[2] == gsa_0123);
+      REQUIRE(vdofs.Max() < NE*NG*NA*ND);
+   }
+}
+
+TEST_CASE("MDGridFunction reshapes", "[MDSpan], [MDGridFunction]")
+{
+   auto is_equal = [&](const  Vector &a, const  Vector &b)
+   {
+      REQUIRE(a.Size() == b.Size());
+      for (int i = 0; i < a.Size(); i++)
+      {
+         const double va = a.GetData()[i], vb =  b.GetData()[i];
+         REQUIRE(va == MFEM_Approx(vb));
+      };
+      return true;
+   };
+
+   SECTION("MDReshapes")
+   {
+      constexpr int p = 2;
+      constexpr int dim = 3;
+      constexpr int nx = 5, ny = 3, nz = 2;
+      Mesh mesh = Mesh::MakeCartesian3D(nx, ny, nz, mfem::Element::HEXAHEDRON);
+
+      H1_FECollection fec_mesh(p, dim);
+      FiniteElementSpace fes_mesh(&mesh, &fec_mesh, dim);
+      mesh.SetNodalFESpace(&fes_mesh);
+
+      L2_FECollection fec(p, dim);
+      FiniteElementSpace fes(&mesh, &fec);
+
+      const std::list<MDLayout<3>> layouts =
+      { {0,1,2}, {0,2,1}, {1,0,2}, {1,2,0}, {2,1,0}, {2,0,1} };
+
+      for (auto &layout: layouts)
+      {
+         constexpr int numGroups = 4, numAngles = 7;
+
+         MDGridFunction<3> psi(&fes, numGroups, numAngles);
+         psi.SetLayout(MDLayout<3>(layout));
+
+         const GridFunction *nodes = mesh.GetNodes();
+         const FiniteElementSpace *mfes = mesh.GetNodalFESpace();
+         const int ng = numGroups, na = numAngles, ne = mfes->GetNE();
+         const ElementDofOrdering e_ordering = ElementDofOrdering::LEXICOGRAPHIC;
+         const Operator *R = mfes->GetElementRestriction(e_ordering);
+         REQUIRE(R);
+         const FiniteElement *mfe = mfes->GetFE(0);
+         const int nd = mfe->GetDof(), vdim = mfes->GetVDim();
+         Vector nodes_e(vdim*nd*ne); nodes_e.UseDevice(true);
+         constexpr int D1D = p + 1;
+         REQUIRE(fes.GetVSize() == D1D*D1D*D1D*ne);
+         nodes_e.Read();
+         REQUIRE(nodes);
+         R->Mult(*nodes, nodes_e);
+         const auto X = mfem::Reshape(nodes_e.Read(), D1D, D1D, D1D, vdim, ne);
+         auto dY = psi.MDWrite();
+
+         MDGridFunction<3> rY1(&fes, numGroups, numAngles);
+         rY1.SetLayout(MDLayout<3>(layout));
```

Comment:
Do you need to set the layout after the constructor?  Could it be part of the constructor?  Or is this just for testing the different layouts?

Reply:
By default, the class's layout (which is part of the template parameters of the constructor) is used. This is more for testing different combinations and could be directly set with `psi.SetLayout(layout);`.